src/main/java/com/spotify/fmt/ForkingExecutor.java

```diff
@@ -0,0 +1,395 @@
+  }
+
+  ForkingExecutor javaArgs(List<String> javaArgs) {
+    this.javaArgs = new ArrayList<>(javaArgs);
+    return this;
+  }
+
+  ForkingExecutor classpath(Collection<String> classpath) {
+    this.configuredClasspath = new ArrayList<>(classpath);
+    return this;
+  }
+
+  ForkingExecutor withDefaultClasspath(boolean withDefaultClasspath) {
+    this.withDefaultClasspath = withDefaultClasspath;
+    return this;
+  }
+
+  private List<String> defaultClasspath() {
+    return Arrays.asList(System.getProperty("java.class.path").split(File.pathSeparator));
+  }
+
+  private List<String> excutionClassPath() {
+    return withDefaultClasspath
+        ? Stream.concat(configuredClasspath.stream(), defaultClasspath().stream())
+            .collect(Collectors.toList())
+        : configuredClasspath;
+  }
+
+  /**
+   * Execute a function in a sub-process.
+   *
+   * @param f The function to execute.
+   * @return The return value of the function. Any exception thrown by the function the will be
+   *     propagated and re-thrown.
+   * @throws IOException if
+   */
+  <T> T execute(SerializableCallable<T> f) throws IOException {
+    try (final Execution<T> execution = new Execution<>(excutionClassPath(), f)) {
+      executions.add(execution);
+      execution.start();
+      return execution.waitFor();
+    }
+  }
+
+  @Override
+  public void close() {
+    executions.forEach(Execution::close);
+  }
+
+  private class Execution<T> implements Closeable {
+
+    private final ExecutorService executor = Executors.newCachedThreadPool();
+
+    private final Path tempdir = Files.createTempDirectory("fmt-maven-plugin");
+
+    private final Path workdir = Files.createDirectory(tempdir.resolve("workdir"));
+
+    private final Path closureFile = tempdir.resolve("closure");
+    private final Path resultFile = tempdir.resolve("result");
+    private final Path errorFile = tempdir.resolve("error");
+
+    private final String home = System.getProperty("java.home");
+    private final Path java = Paths.get(home, "bin", "java").toAbsolutePath().normalize();
+
+    private final List<String> classpath;
+
+    private final SerializableCallable<T> f;
+
+    private Process process;
+
+    Execution(List<String> classpath, SerializableCallable<T> f) throws IOException {
+      this.classpath = classpath;
+      this.f = Objects.requireNonNull(f);
+    }
+
+    void start() {
+      if (process != null) {
+        throw new IllegalStateException();
+      }
+      log.debug("serializing closure");
+      try {
+        Serialization.serialize(f, closureFile);
+      } catch (SerializationException e) {
+        throw new RuntimeException("Failed to serialize closure", e);
+      }
+
+      final String classPathArg = String.join(File.pathSeparator, classpath);
+
+      final ProcessBuilder processBuilder =
+          new ProcessBuilder(java.toString(), "-cp", classPathArg).directory(workdir.toFile());
+
+      // Propagate -Xmx and -D.
```

Comment:
Why not any -XX etc?

Reply:
Good question. That’s a remnant from the flo use case. If this was invoking google-java-format as a cli then I’d be naively leaning towards not propagating any flags at all. 

But maybe it’d useful to propagate all flags? Not sure 🤔

Reply:
Propagating Xmx as-is might even be problematic if it’s high.

Reply:
I wonder what surefire etc does.

Reply:
The surefire docs are confusing on this point but having stepped through the surefire execution of forks with a debugger and having inspected the precise command line and serialized property files etc I feel confident in saying that surefire does not automatically propagate any properties, not even `-Xmx` etc. One has to explicitly allow-list specific properties to propagate. 

Maybe we should consider conservatively not propagating properties here unless we see a specific need for it in google-java-format.